### PR TITLE
transformations: Convert `memref` to `ptr`dialect

### PR DESCRIPTION
**Pull request by mamanain**
Introducing lowering of some memref operations to a ptr dialect. This will be used to translate riscv compilation to ptr dialect in the future (this is the second pr in the series).

### CODE REVIEW

File: xdsl/transforms/convert_memref_to_ptr.py [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+    memref_type: memref.MemRefType[Any],
+    indices: Iterable[SSAValue],
+) -> tuple[list[Operation], SSAValue]:
+    ops: list[Operation] = [
+        memref_ptr := ptr.ToPtrOp(
+            operands=[target_memref], result_types=[ptr.PtrType()]
+        )
+    ]
+
+    if len(list(indices)) == 0:
```
Comment (superlopuh) [inline xdsl/transforms/convert_memref_to_ptr.py:108]:
```suggestion
    if not indices:
```

Comment (superlopuh) [reply xdsl/transforms/convert_memref_to_ptr.py:108]: It looks like the indices passed below are actually a Collection, so this should be possible, and would avoid an extra list construction

File: xdsl/transforms/convert_memref_to_ptr.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+
+
+@dataclass(frozen=True)
+class ConvertMemrefToPtr(ModulePass):
+    name = "convert-memref-to-ptr"
+
+    def apply(self, ctx: MLContext, op: builtin.ModuleOp) -> None:
+        the_one_pass = PatternRewriteWalker(
+            GreedyRewritePatternApplier([ConvertStoreOp(), ConvertLoadOp()]),
+            apply_recursively=False,
```
Comment (superlopuh) [inline xdsl/transforms/convert_memref_to_ptr.py:160]: Is this necessary?

Comment (superlopuh) [reply xdsl/transforms/convert_memref_to_ptr.py:160]: Similarly for the other options below

File: xdsl/transforms/convert_memref_to_ptr.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+    GreedyRewritePatternApplier,
+    PatternRewriter,
+    PatternRewriteWalker,
+    RewritePattern,
+    op_type_rewrite_pattern,
+)
+from xdsl.utils.exceptions import DiagnosticException
+
+
+# I think we also need to pass the adress width.
```
Comment (superlopuh) [inline xdsl/transforms/convert_memref_to_ptr.py:20]: Isn't it computed below?

**Review by superlopuh — APPROVED** [review]
Could you please add some doc strings to the helpers? LGTM otherwise

File: xdsl/transforms/convert_memref_to_ptr.py (hunk 3 above)
Comment (mamanain) [reply xdsl/transforms/convert_memref_to_ptr.py:20]: yeah, forgot to delete this comment from a previous version

File: xdsl/transforms/convert_memref_to_ptr.py [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+
+    assert isinstance(memref_type.element_type, builtin.FixedBitwidthType)
+
+    match memref_type.layout:
+        case builtin.NoneAttr():
+            strides = builtin.ShapedType.strides_for_shape(memref_type.get_shape())
+        case builtin.StridedLayoutAttr():
+            strides = memref_type.layout.get_strides()
+        case _:
+            raise DiagnosticException(f"Unsupported layout type {memref_type.layout}")
```
Comment (superlopuh) [inline xdsl/transforms/convert_memref_to_ptr.py:33]: Oh I forgot about this, could you please add a small PR with a test that catches this diagnostic exception?

**Pull request merged**